diff --git a/.gitignore b/.gitignore
index 5c33438..b49f865 100644
--- a/.gitignore
+++ b/.gitignore
@@ -23,3 +23,4 @@ ph
 barrier
 /lab-*.json
 .DS_Store
+.vscode
\ No newline at end of file
diff --git a/Makefile b/Makefile
index 365c91b..3c6338b 100644
--- a/Makefile
+++ b/Makefile
@@ -188,6 +188,7 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
+	$U/_sleep\
 
 
 
diff --git a/user/sleep.c b/user/sleep.c
new file mode 100644
index 0000000..4635193
--- /dev/null
+++ b/user/sleep.c
@@ -0,0 +1,25 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "kernel/fcntl.h"
+#include "user/user.h"
+
+int
+main(int argc, char* argv[])
+{
+    if (argc != 2) // incorrect number of arguments
+    {
+        fprintf(2, "Usage: sleep ticks\n");
+        exit(1);
+    }
+
+    int n_ticks = atoi(argv[1]); // convert tick argument to int
+    if (n_ticks <= 0) // avoid non-positive number of ticks
+    {
+        fprintf(2, "Error: Invalid number of ticks\n");
+        exit(1);
+    }
+
+    sleep(n_ticks); // call sleep from user.h
+
+    exit(0); // sleep success
+}
\ No newline at end of file
